### DISCUSSION TITLE
Modelling token sets

### DISCUSSION THREAD

**Discussion by uuf6429** [antlr/antlr4#4350] [Q&A] [answered]
I've recently started using Antlr to build a prototype of a language I have in mind.
One of the interesting functionalities is the ability to switch between sets of tokens.
For example, a "traditional set" would look like:
```antlr
PROP_ACCESS: '.';
STMT_TERM:   ';';
MTD_L:       '(';
MTD_R:      '(';
```
Whereas a "natural set" could look like:
```antlr
PROP_ACCESS: ' ';
STMT_TERM:   '.';
MTD_L:      ' ';
MTD_R:      ' ';
```

My first thought was to somehow switch between these sets based on some top-level config/pragma/whatever (is this even possible?). Or perhaps it makes more sense to actually have separate grammars?

**Comment by ericvergnaud** [comment, marked as answer]
You’d definitely want separate grammars, but FYI the latter is not valid i.e. 3 of the tokens have the same pattern so only 1 will ever be generated Envoyé de mon iPhoneLe 8 juil. 2023 à 12:26, Christian Sciberras ***@***.***> a écrit :﻿
I've recently started using Antlr to build a prototype of a language I have in mind.
One of the interesting functionalities is the ability to switch between sets of tokens.
For example, a "traditional set" would look like:
PROP_ACCESS: '.';
STMT_TERM:   ';';
MTD_L:       '(';
MTD_R:      '(';
Whereas a "natural set" could look like:
PROP_ACCESS: ' ';
STMT_TERM:   '.';
MTD_L:      ' ';
MTD_R:      ' ';
My first thought was to somehow switch between these sets based on some top-level config/pragma/whatever (is this even possible?). Or perhaps it makes more sense to actually have separate grammars?

**Reply by uuf6429** [reply]
> but FYI the latter is not valid

I suspected that as well, and was wondering what would happen.

Anyway, thanks for the reply!

PS: I was able to jump into the subject and do a lot of stuff with the antlr jetbrains plugin - that's very cool and the effort well appreciated!

**Reply by KvanTTT** [reply]
> You’d definitely want separate grammars, but FYI the latter is not valid i.e. 3 of the tokens have the same pattern so only 1 will ever be generated

Not correct. All three tokens will be generated but [TOKEN_UNREACHABLE](https://github.com/antlr/antlr4/blob/dev/tool/src/org/antlr/v4/tool/ErrorType.java#L1080) warnings will be reported.

**Comment by ericvergnaud** [comment]
I meant they won’t be generated at runtimeEnvoyé de mon iPhoneLe 8 juil. 2023 à 16:20, Ivan Kochurkin ***@***.***> a écrit :﻿

You’d definitely want separate grammars, but FYI the latter is not valid i.e. 3 of the tokens have the same pattern so only 1 will ever be generated

Not correct. All three tokens will be generated but TOKEN_UNREACHABLE warnings will be reported.

**Reply by KvanTTT** [reply]
Ok, in runtime they won't be *recognized* but not *generated* (using term `generation` for both generation and runtime sounds confusing).
